### PR TITLE
Record performance statistics during data loading

### PR DESCRIPTION
**Pull request by Zhangshunyu**
1. After data loaded successfully, the main statistics information will be printed by each executor
2. At the same time, some tiny info like write one temp sort file/compress mdk data time/write block.. time will be printed immediately in log.

I just take CarbonExample who loads only 1000 lines data as an example, you can see the statistics info as following in executor side:
```
]==================== DummyLoadRDD Cost Time: 5.085(s)
]============================== lru Cache Load Cost Time: 0.024(s)
]==========**********Compress One Node For One Thread pool-24-thread-1 Cost: 0.026
]=====##### Write One Node Cost Time: 0.0#####=====
]========== TIME STATISTICS PartitionID: 0==========
]STATISTICS ->Raw data IO cost(load csv to dataFrame and generate block dict): 0.132(s)
]STATISTICS ->Distinct Value IO cost(global dict shuffle and write dict file): 0.648(s)
]STATISTICS ->  |_maximum distinct column shuffle time: 0.162(s)
]STATISTICS ->  |_maximum distinct column write dict file time: 0.35(s)
]STATISTICS ->Total cost of gen surrogate key, sort and write to temp files: 0.522(s)
]STATISTICS ->  |_read cost of raw csv file: 0.305(s)
]STATISTICS ->  |_cost of transform to surrogate key: 0.485(s)
]STATISTICS ->  |_io cost(sort rows and write to temp file): 0.417(s)
]STATISTICS ->IO cost(tansform to MDK, compress and write fact file): 0.629(s)
]==============================
]========== BLOCK_INFO ==========
]BLOCK_INFO ->Node host: localhost
]BLOCK_INFO ->The block count in this node: 1
]==============================
```

### CODE REVIEW

**Comment by jackylk** [comment]
can you take a look a #356  to check whether you can leverage the Profiler there?

**Comment by Zhangshunyu** [comment]
@This pr has no impact on pr 356.

**Comment by Zhangshunyu** [comment]
@jackylk This pr has no impact on pr 356.

**Comment by jackylk** [comment]
I mean can you re-use the profiler in pr356. pr356 has a profiler that you can define your load phase. We can use the same to record performance for both query and load

**Comment by kumarvishal09** [comment]
@Zhangshunyu please check the performance impact after you change

File: core/src/main/java/org/carbondata/core/util/CarbonTimeStatisticsUtil.java [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+          Collections.synchronizedSortedSet(new TreeSet<Long>());
+  //Due to thread thread blocking in each task, we only record the max
+  //dicWriteFile Time of each single thread
+  private static SortedSet<Long> dicWriteFileMaxTime =
+          Collections.synchronizedSortedSet(new TreeSet<Long>());
+  private static SortedSet<Long> dicShuffleAndWriteFileTotalTime =
+          Collections.synchronizedSortedSet(new TreeSet<Long>());
+
+  //Generate surrogate keys total time:
+  private static ConcurrentHashMap<String, SortedSet<Long>> parSurrogatekeysTotalTimeMap =
```
Comment (kumarvishal09) [inline core/src/main/java/org/carbondata/core/util/CarbonTimeStatisticsUtil.java]: private static Map<String, SortedSet<Long>> parCsvInputStepTimeMap =
 +          new ConcurrentHashMap<String, SortedSet<Long>>();

**Comment by kumarvishal09** [comment]
for data loading and query statistics we can have one common interface, which will have multiple implementation, default will be dummy which will not record any thing . And one property we can expose like carbon.record.statistics. factory will return the concrete implementation based on the property value. in normal case it will record dummy so it will not impact the performance

File: core/src/main/java/org/carbondata/core/util/CarbonTimeStatisticsUtil.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.carbondata.common.logging.LogService;
+import org.carbondata.common.logging.LogServiceFactory;
+
+/**
+ * A util which provide a method used to count time information.
+ */
+public class CarbonTimeStatisticsUtil {
```
Comment (kumarvishal09) [inline core/src/main/java/org/carbondata/core/util/CarbonTimeStatisticsUtil.java]: there are so many map and set used in this class, why this is required?

File: core/src/main/java/org/carbondata/core/util/CarbonTimeStatisticsUtil.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+  private static final LogService LOGGER =
+          LogServiceFactory.getLogService(CarbonTimeStatisticsUtil.class.getName());
+  //Generate global dictionay total time :
+  private static SortedSet<Long> loadCsvfilesToDfTime =
```
Comment (kumarvishal09) [inline core/src/main/java/org/carbondata/core/util/CarbonTimeStatisticsUtil.java]: how many values we are going to store here in all the map and sets, i think only max value,then in that case we can have one class which will be used to store the max value .

**Pull request closed without merging**